IPIP/0000-double-hash-dht.md

```diff
@@ -0,0 +1,339 @@
+5. Content Provider takes the current timestamp `TS`.
+6. Content Provider signs `EncPeerID` and `TS` using its private key. `Signature = Sign(privkey, EncPeerID || TS)`
+7. Content Provider computes `ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)`.
+8. Once the lookup request has returned the 20 closest peers, Content Provider sends a Publish request to these DHT servers. The Publish request contains [`HASH2`, `EncPeerID`, `TS`, `Signature`, `ServerKey`].<!-- TODO: define exact format -->
+9. Each DHT server verifies `Signature` against the `PeerID` of the Content Provider used to open the libp2p connection. `Verify(CPPeerID, Signature, EncPeerID || TS)`. It verifies that `TS` is _recent enough_. If invalid, send an error to the client. <!-- TODO: define error && check TS valid -->
+10. Each DHT server adds an entry in their Provider Store for `HASH2` -> `ServerKey` -> `CPPeerID` -> [`EncPeerID`, `TS`, `Signature`], with `CPPeerID` being the `PeerID` of the Content Provider. If there is already an entry including `CPPeerID` for `HASH2` -> `ServerKey`, and if the `TS` of the new valid entry is newer than the existing `TS`, overwrite the entry in the Provider Store. Else drop the new entry.
+11. Each DHT server confirms to Content Provider that the Provider Record has been successfully added.
+12. The proces is over once Content Provider has received 20 confirmations.
+
+**Lookup Process**
+1. Client computes `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)` (`MH` is the MultiHash included in the CID).
+2. Client selects a prefix of `HASH2`, `KeyPrefix = HASH2[:l]` for a defined `l` (see [`l` selection](#prefix-length-selection)).
+2. Client finds the closest `PeerID`s to `HASH2` in XOR distance in its Routing Table.
+3. Client sends a DHT lookup request for `KeyPrefix` to these DHT servers. The request contains a flag to specify whether Client wants the `multiaddrs` associated with the `CPPeerID` or not.
+4. The DHT servers find the 20 closest `PeerID`s to `KeyPrefix` in XOR distance (see [algorithm](#closest-keys-to-a-key-prefix)). Add these `PeerID`s and their associated multiaddresses (if applicable) to the `message` that will be returned to Client.
+5. The DHT servers search if there are entries matching `KeyPrefix` in their Provider Store.
+6. For all entries `HASH2` of the Provider Store where `HASH2[:len(KeyPrefix)]==KeyPrefix`, add to `message` the following encrypted payload: `Enc(ServerKey, EncPeerID || TS || Signature || multiaddrs)`, for `multiaddrs` being the multiaddresses associated with `CPPeerID` (if applicable) if the `multiaddrs` were requested by Client. DHT Servers can decide to put a maximal limit of returned Provider Record per request. If too many `HASH2` are matching `KeyPrefix`, they select randomly 128 matching provider records per request, and send a flag to Client to signal that the limit was reached. <!-- TODO: define flag-->
```

Comment:
What is the reason behind using a second layer of encryption here: `Enc(ServerKey, EncPeerID || TS || Signature || multiaddrs)`?

Reply:
Sending `EncPeerID || TS || Signature` in the clear would enable any peer replaying the attack without knowing `CID` to learn about the Content Provider. It is easy to crawl the IPFS network and get the `PeerID`s of all nodes, and then try to verify the `Signature` against all `PeerID`s (bruteforce). The verification only succeeds for the Content Provider.

Moreover, this extra encryption layer allows to send the Content Providers' `multiaddrs` without leaking any information to node potentially replaying the DHT request, while not knowing `CID`. If DHT Server returns a valid `multiaddrs`, 1 additional DHT lookup is spared during content resolution, speeding it up ~2x. The impact is expected to be significant when https://github.com/libp2p/go-libp2p-kad-dht/pull/795 gets merged, as the `multiaddrs` TTL will be increasing.

References for `multiaddrs` TTL:
- [RFM17.1](https://github.com/protocol/network-measurements/blob/master/results/rfm17.1-sharing-prs-with-multiaddresses.md)
- https://github.com/ipfs/kubo/issues/9264
- https://github.com/libp2p/go-libp2p-kad-dht/pull/795

Reply:
Got it, thanks for explaining. Maybe we can separate up authorization data from the payload itself? As different content routing systems might use differnet authorization mechanics. What I'm proposing is this:

`EncPeerID || Enc(ServerKey, TS || multiaddrs)`

Reply:
@ischasny you didn't include the `Signature`. As the `Signature` must be protected by encryption, I suggest the following:

`EncPeerID || Enc(ServerKey, TS || Signature || multiaddrs)`

Would this be better for the Indexers?

Reply:
Perfect, thanks @guillaumemichel!